bundles/org.openhab.binding.flume/src/main/java/org/openhab/binding/flume/internal/api/FlumeApi.java

```diff
@@ -0,0 +1,411 @@
+    public static final String API_UNITS_METRIC = "LITERS";
+
+    // @formatter:off
+    public enum OperationType {
+        SUM, AVG, MIN, MAX, CNT;
+
+        public static boolean contains(String value) {
+            return Arrays.stream(values()).anyMatch((t) -> t.name().equals(value));
+        }
+    }
+
+    public enum BucketType {
+        YR, MON, DAY, HR, MIN;
+
+        public static boolean contains(String value) {
+            return Arrays.stream(values()).anyMatch((t) -> t.name().equals(value));
+        }
+    }
+
+    public enum SortDirectionType {
+        ASC, DESC
+    }
+    // @formatter:on
+
+    protected String clientId = "";
+    protected String clientSecret = "";
+    protected String username = "";
+    protected String password = "";
+    protected Gson gson;
+
+    private String accessToken = "";
+    private String refreshToken = "";
+    private int userId;
+    private LocalDateTime tokenExpiresAt = LocalDateTime.now();
+
+    private HttpClient httpClient;
+
+    public FlumeApi(HttpClient httpClient) {
+        this.httpClient = httpClient;
+        this.gson = new GsonBuilder()
+                .registerTypeAdapter(LocalDateTime.class, new JsonLocalDateTimeSerializer("yyyy-MM-dd HH:mm:ss")) // 2022-07-13
+                                                                                                                  // 20:14:00
+                .registerTypeAdapter(ZonedDateTime.class, new JsonZonedDateTimeSerializer()) // 2022-07-14T03:13:00.000Z
+                .create();
+    }
+
+    public String getClientId() {
+        return clientId;
+    }
+
+    public void initialize(String clientId, String clientSecret, String username, String password, ThingUID bridgeUID)
+            throws FlumeApiException, IOException, InterruptedException, TimeoutException, ExecutionException {
+        this.clientId = clientId;
+        this.clientSecret = clientSecret;
+        this.username = username;
+        this.password = password;
+
+        getToken();
+    }
+
+    private void getToken() throws FlumeApiException, IOException, InterruptedException, TimeoutException,
+            ExecutionException, NullPointerException {
+        FlumeApiGetToken getToken = new FlumeApiGetToken();
+
+        getToken.clientId = clientId;
+        getToken.clientSecret = clientSecret;
+        getToken.username = username;
+        getToken.password = password;
+
+        String url = APIURL_BASE + APIURL_TOKEN;
+        Request request = httpClient.newRequest(url).method(HttpMethod.POST)
+                .header(HttpHeader.CONTENT_TYPE, "application/json")
+                .content(new StringContentProvider(gson.toJson(getToken)), "application/json");
+
+        logger.trace("POST: {}", url);
+        ContentResponse response = request.send();
+
+        if (response.getStatus() == 400) {
+            throw new FlumeApiException("@text/api.invalid-user-credentials [\"" + response.getReason() + "\"]",
```

Comment:
400 (bad request) seems to be translated to `api.invalid-user-credentials` I would assume 401 is for invalid credentials/not authenticated.

Reply:
I just tested this again and it does send 400 for invalid credentials.  I added a comment to the code stating this.